src/opendr/perception/object_detection_2d/gem/gem_learner.py

```diff
@@ -0,0 +1,1059 @@
+        if m1_edataset is None:
+            m1_edataset = ExternalDataset(dataset_location, 'coco')
+        if m2_edataset is None:
+            m2_edataset = ExternalDataset(dataset_location, 'coco')
+
+        if annotations_folder is None:
+            annotations_folder = self.datasetargs.annotations_folder
+        if m1_annotations_file is None:
+            m1_annotations_file = self.datasetargs.m1_val_annotations_file
+        if m2_annotations_file is None:
+            m2_annotations_file = self.datasetargs.m2_val_annotations_file
+        if m1_images_folder is None:
+            m1_images_folder = self.datasetargs.m1_val_images_folder
+        if m2_images_folder is None:
+            m2_images_folder = self.datasetargs.m2_val_images_folder
+
+        if self.postprocessors is None:
+            self.__create_postprocessors()
+
+        self.__create_criterion()
+
+        device = torch.device(self.device)
+
+        m1_dataset = self.__prepare_dataset(
+            m1_edataset,
+            image_set="m1_val",
+            images_folder_name=m1_images_folder,
+            annotations_folder_name=annotations_folder,
+            annotations_file_name=m1_annotations_file
+        )
+        m1_sampler_main = SequentialSampler(m1_dataset)
+        m1_eval_seed = m1_dataset.set_seed()
+
+        m2_dataset = self.__prepare_dataset(
+            m2_edataset,
+            image_set="m2_val",
+            images_folder_name=m2_images_folder,
+            annotations_folder_name=annotations_folder,
+            annotations_file_name=m2_annotations_file,
+            seed=m1_eval_seed
+        )
+        m2_sampler_main = SequentialSampler(m2_dataset)
+
+        if not self.args.distributed:
+            m1_sampler = m1_sampler_main
+            m2_sampler = m2_sampler_main
+        else:
+            m1_sampler = DistributedSamplerWrapper(m1_sampler_main)
+            m2_sampler = DistributedSamplerWrapper(m2_sampler_main)
+
+        m1_data_loader = DataLoader(
+            m1_dataset,
+            self.batch_size,
+            sampler=m1_sampler,
+            drop_last=False,
+            collate_fn=utils.collate_fn,
+            num_workers=self.args.num_workers
+        )
+        m2_data_loader = DataLoader(
+            m2_dataset,
+            self.batch_size,
+            sampler=m2_sampler,
+            drop_last=False,
+            collate_fn=utils.collate_fn,
+            num_workers=self.args.num_workers
+        )
+
+        if isinstance(m1_edataset, ExternalDataset):
+            base_ds = get_coco_api_from_dataset(m1_dataset)
+        else:
+            base_ds = None
+
+        test_stats, _ = evaluate(
+            self.model, self.criterion, self.postprocessors,
+            m1_data_loader, m2_data_loader, base_ds, device,
+            self.temp_path, verbose=verbose
+        )
+
+        return test_stats
+
+    def infer(self, m1_image, m2_image):
```

Comment:
Why this function receive two images as input?

Reply:
The Gem algorithm is a multimodal 2D object detection algorithm. It therefore requires two inputs, e.g. an RGB and an infrared image. I now see that the docstrings should be updated. I will do this and I hope that it is clear then.